jxl/src/frame/modular/transforms/rct.rs

```diff
@@ -0,0 +1,88 @@
+// Copyright (c) the JPEG XL Project Authors. All rights reserved.
+//
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+use crate::{
+    error::Result,
+    frame::modular::{
+        transforms::{RctOp, RctPermutation},
+        ModularBufferInfo,
+    },
+    image::Image,
+    util::tracing_wrappers::*,
+};
+
+use super::TransformStep;
+
+#[instrument(skip(buffers), ret)]
+pub fn do_rct_step(
+    step: TransformStep,
+    buffers: &mut [ModularBufferInfo],
+    (gx, gy): (usize, usize),
+) -> Result<Vec<(usize, usize)>> {
+    let TransformStep::Rct {
+        buf_in,
+        buf_out,
+        op,
+        perm,
+    } = step
+    else {
+        unreachable!("asking to RCT a non-RCT step");
+    };
+
+    let grid = buffers[buf_in[0]].grid_shape.0 * gy + gx;
+    for i in 0..3 {
+        assert_eq!(buffers[buf_in[0]].grid_kind, buffers[buf_in[i]].grid_kind);
+        assert_eq!(buffers[buf_in[0]].grid_kind, buffers[buf_out[i]].grid_kind);
+        assert_eq!(buffers[buf_in[0]].info.size, buffers[buf_in[i]].info.size);
+        assert_eq!(buffers[buf_in[0]].info.size, buffers[buf_out[i]].info.size);
+        assert!(buffers[buf_in[i]].buffer_grid[grid].data.is_some());
+        assert!(buffers[buf_out[i]].buffer_grid[grid].data.is_none());
+    }
+    if op != RctOp::YCoCg && perm != RctPermutation::Rgb {
+        unimplemented!("non-YCoCg not implemented");
```

Comment:
the condition and the error message don't quite agree. Should this be 
"non-YCoCg with non rgb perm not implemented" or something?

Reply:
it could be, but I'm not sure it's _too_ important...